blog/_posts/2022-12-13-tut-access-to-public-ip-from-vms-containers-using-firewalld.md

```diff
@@ -0,0 +1,52 @@
+---
+layout: post
+title: "Access to public IP from VMs/Containers using Firewalld"
+section: Blog
+date: 2022-12-13T13:00:00
+author: David Foley
+category: tutorial
+---
+
+You are running some LXC containers on a host and you use Firewalld to forward ports from the public internet to the containers. How do you enable access to the public IP of the LXD host from the LXC containers?
+
+## TL;DR
+- Ordinairly connections from containers to to services which are reached by port forwarding on Firewalld fail, because after outputting the packet to the public interface, it never returns and therefore Firewalld cannot process the port forward.
+- The soloution is destination NAT: `sudo firewall-cmd --zone=trusted --add-rich-rule='rule family="ipv4" destination address="123.123.123.123" forward-port port="80" protocol="tcp" to-port="80" to-addr="10.10.1.20"'`
+
+## The Scenario
+You have some LXC containers running on a host, the default LXD setup creates a virtual bridge to which all the containers are connected, they have their own private network say in the 10.10.1.0/24 subnet.
+
+You use Firewalld to forward ports from the public internet to the containers. In this scenario when a container does a DNS lookup to which the answer is the public IP address of the LXD host and the container then tries to connect to say port 80 on that public IP it will fail. Why? The HTTP request is received on the input chain by firewalld, no processing is required as it appears to be destined for the public internet. Firewalld outputs the packet to the public interface of the host. Since the HTTP Proxy is not bound to the public interface, it is instead reached via port forwards in Firewalld, the connection fails. This is because Firewalld handles the port forwarding and, after outputting the packet to the public interface, it never returns and therefore Firewalld cannot process the port forward.
```

Comment:
> HTTP request is received on the input chain by firewalld, no processing is required as it appears to be destined for the public internet.

You are correct that it's received on INPUT, but that means it's destined to the LXC _HOST_. Not the internet.


> Since the HTTP Proxy is not bound to the public interface, it is instead reached via port forwards in Firewalld, the connection fails.

Rearranging the wording would help here.
"Since the HTTP Proxy is not bound to the public interface the connection fails. It instead must be reached via port forwards in Firewalld."

Reply:
Apologies, the first point was phrased poorly. I was trying to convey that it "appears" from the Firewalld perspective that it is destined to be OUTPUT, no further processing will be done. 

In any case I am happy to correct it. I will also amend the second point I agree my version is clunky.

Thank you